### PR TITLE
Replace transaction set in ConsensusData with hashes

### LINKED ISSUE #993: Propagate newly received transactions via ConsensusData to the TransactionPool
Currently, when we receive transactions via `ConsensusData` we do not store the transaction in the pool.
It would be wise to do so, and could also allow us to find out some issues (e.g. if we're stuck trying to nominate conflicting values).

Definition of done:
- [ ] When a node with an empty pool receive `ConsensusData` that contains transactions it is not aware of, it stores it in its `TransactionPool`, even if it does not reach consensus;
- [ ] A node does not store transactions which are conflicting;
- [ ] If a transaction conflicts, or was newly added, it is reported in the log (as debug);

### PR DESCRIPTION
**Pull request by omerfirmak**
Part of #1228

Fixes https://github.com/bpfkorea/agora/issues/993

### CODE REVIEW

File: source/agora/node/FullNode.d
Comment (AndrejMitrovic) [inline source/agora/node/FullNode.d]: That's not what the code does though, it doesn't return `Transaction.init` but just skips those transactions entirely.

File: source/agora/node/FullNode.d
Comment (AndrejMitrovic) [inline source/agora/node/FullNode.d:790]: I feel like we can optimize this in the future. I am pretty sure databases are well optimized for looking up several keys at once, and not just one at a time.

File: source/agora/consensus/protocol/Nominator.d
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Nominator.d]: By SCP's definition if you return `kInvalidValue` then it means the node may never vote this value as valid afterwards. Are you sure we're not supposed to use `kMaybeValidValue`?

Comment (omerfirmak) [reply source/agora/consensus/protocol/Nominator.d]: Well, initially we returned `kMaybeValidValue` here. But I had some problems in the balloting phase. Node started to confirm statements without having access to all the previous blocks. I completely went from these [two](https://github.com/bpfkorea/agora/blob/bbf523199b6efd1667ee0e3d79ad59d4c3ea93c2/source/agora/consensus/protocol/Nominator.d#L539) [comments](https://github.com/bpfkorea/agora/blob/406a22df4c33dfd2f2e3a61811fe6e73aea94852/source/scpp/src/scp/SCPDriver.h#L56-L58). and decided `kInvalidValue` was more appropriate. In nominating phase it does not matter what we return they are the same. In balloting we should still fully validate the value, but our ledger may be a bit behind to return `kMaybeValidValue`. That is indeed what Stellar does actually. They only return `kMaybeValidValue` [here](https://github.com/stellar/stellar-core/blob/54b03f755ae5d5aa12a799c8f1ee4d87fc9d1a1d/src/herder/HerderSCPDriver.cpp#L237), which is the case where the value is fully validated but ledger is not fully caught up.

File: source/agora/node/FullNode.d
Comment (omerfirmak) [reply source/agora/node/FullNode.d:790]: Agreed, this was just to get stuff working.

File: source/agora/node/FullNode.d
Comment (omerfirmak) [reply source/agora/node/FullNode.d]: Oops, copy paste error.

File: source/agora/consensus/protocol/Nominator.d
Comment (omerfirmak) [reply source/agora/consensus/protocol/Nominator.d]: > then it means the node may never vote this value as valid afterwards

I am not entirely sure, but if the same value is renominated, it should probably count as a new value.

Comment (AndrejMitrovic) [reply source/agora/consensus/protocol/Nominator.d]: Well okay. The docs are not amazing..

Comment (AndrejMitrovic) [reply source/agora/consensus/protocol/Nominator.d]: I mean the SCP docs.

File: source/agora/node/Ledger.d
Comment (AndrejMitrovic) [inline source/agora/node/Ledger.d]: Is "ctx" supposed to mean "context"?

File: source/agora/network/NetworkManager.d
Comment (AndrejMitrovic) [inline source/agora/network/NetworkManager.d:887]: You need to check the return value.

**Review by AndrejMitrovic — COMMENTED** [review]
Overall LGTM. Just a few issues.

File: source/agora/network/NetworkManager.d
Comment (omerfirmak) [reply source/agora/network/NetworkManager.d:887]: Do we want to log it, if its invalid? I could not think of any other reason for checking if it is valid or not.

File: source/agora/node/Ledger.d
Comment (omerfirmak) [reply source/agora/node/Ledger.d]: Yep

File: source/agora/network/NetworkManager.d
Comment (AndrejMitrovic) [reply source/agora/network/NetworkManager.d:887]: You can't think of any? Check the next line.

If the tx is invalid then the missing tx set should not be modified..

Comment (AndrejMitrovic) [reply source/agora/network/NetworkManager.d:887]: Wait I just realized something.

Comment (AndrejMitrovic) [reply source/agora/network/NetworkManager.d:887]: My assumption was wrong. In fact if the tx is invalid there is no other tx with a matching hash, that would make it a hash collision.

Comment (AndrejMitrovic) [reply source/agora/network/NetworkManager.d:887]: I wonder if this can be abused in some way. If a bad actor nominates a big tx set of invalid hashes, will that cause the node to constantly try to download these txes?

Comment (AndrejMitrovic) [reply source/agora/network/NetworkManager.d:887]: by invalid I mean fake hashes of non-existing txes.

Comment (omerfirmak) [reply source/agora/network/NetworkManager.d:887]: It was hard to follow but, yes :)

Comment (omerfirmak) [reply source/agora/network/NetworkManager.d:887]: Oops I didnt see the last 2 messages when I replied to you.
 
 > I wonder if this can be abused in some way. If a bad actor nominates a big tx set of invalid hashes, will that cause the node to constantly try to download these txes?
 
Until the end of that round, yes. We clear the unknown txs at the end of each round. "big tx set" will probably be bound by some upper limit too eventually. Maybe the block size or tx count?

Comment (omerfirmak) [reply source/agora/network/NetworkManager.d:887]: We can maybe give up on looking for hashes after a certain amount of tries

Comment (AndrejMitrovic) [reply source/agora/network/NetworkManager.d:887]: Anyway it shouldn't block the PR.

**Comment by omerfirmak** [comment]
I rebased the fee dıstrıbutıon commits and added some changes to better handle the externalization with missing TXs.
Tests pass on my local machine but I feel like block signatures could be easily broken by this. @hewison-chris, can you check if these make sense to you, especially the early returns in the last commit .

The last commit will be formatted/documented and mostly squashed to previous commits.

File: source/agora/consensus/protocol/Nominator.d
Comment (hewison-chris) [inline source/agora/consensus/protocol/Nominator.d:892]: I feel like this should be an assert(0) here as we can not ignore the invalid tx set when we should be externalizing the block.

Comment (omerfirmak) [reply source/agora/consensus/protocol/Nominator.d:892]: Since we now externalize TX hashes, we might not have all the TXs in our pool at the point the value was externalized. So that could fail for missing TXs as well. In that case we should be able to catch up to the network and sign the unsigned blocks along the way.

Comment (hewison-chris) [reply source/agora/consensus/protocol/Nominator.d:892]: So will this externalize be triggered again after the missing TXs are received?

Comment (omerfirmak) [reply source/agora/consensus/protocol/Nominator.d:892]: No, node will have to download the full block from the network.

Comment (hewison-chris) [reply source/agora/consensus/protocol/Nominator.d:892]: Perhaps I should also not assert in that case, when there are not enough signatures. The node can also wait to recieve the block during a periodic catchup and test again for required number of signatures before externalizing.

Comment (hewison-chris) [reply source/agora/consensus/protocol/Nominator.d:892]: I created [PR #1517](https://github.com/bpfkorea/agora/pull/1517) to address this for block sigs. I agree with your logic here.

Comment (omerfirmak) [reply source/agora/consensus/protocol/Nominator.d:892]: Great, thanks!

File: source/agora/consensus/protocol/Nominator.d
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Nominator.d]: So it returns `true` if it fails?

Comment (AndrejMitrovic) [reply source/agora/consensus/protocol/Nominator.d]: Ah wait it returns a string.

I thought this behavior was deprecated.

Comment (AndrejMitrovic) [reply source/agora/consensus/protocol/Nominator.d]: This would make it clearer:

`if (auto error = this.ledger.getValidTXSet(con_data, received_tx_set))`

Although you could argue that the string is unused anyway.

Comment (omerfirmak) [reply source/agora/consensus/protocol/Nominator.d]: Yeah, lets just make it explicit.

Comment (hewison-chris) [reply source/agora/consensus/protocol/Nominator.d]: If the string is not used then it may as well be a bool. It is not an `isInvalid...` function so it is not obvious what is success status.

**Review by AndrejMitrovic — APPROVED** [review]

**Comment by AndrejMitrovic** [comment]
I'll let @hewison-chris finish his review & then we can merge.

File: source/agora/consensus/protocol/Nominator.d
Comment (hewison-chris) [inline source/agora/consensus/protocol/Nominator.d:542]: We probably should log here the fail_reason as there is nothing logged in any of the called functions.

File: source/agora/consensus/protocol/Nominator.d
Comment (hewison-chris) [inline source/agora/consensus/protocol/Nominator.d:892]: Also here

**Review by hewison-chris — COMMENTED** [review]
Looks good, only a couple of logging requests.

**Review by linked0 — APPROVED** [review]
LGTM

**Comment by omerfirmak** [comment]
Thanks for the reviews everyone. I have made the changes and rebased. I think this is good to merge it once it goes green.

**Comment by AndrejMitrovic** [comment]
> Thanks for the reviews everyone. I have made the changes and rebased. I think this is good to merge it once it goes green.

It would be good to test this in our test environment. It's a significant change.

**Comment by omerfirmak** [comment]
> > Thanks for the reviews everyone. I have made the changes and rebased. I think this is good to merge it once it goes green.
> 
> It would be good to test this in our test environment. It's a significant change.

Together with the Coinbase TXs they make a dangerous duo. Once we get `faucet` to include fees we should be able to test extensively.

**Comment by omerfirmak** [comment]
@hewison-chris , is this the failure we were seeing in NetworkManager tests?
https://github.com/bpfkorea/agora/pull/1495/checks?check_run_id=1739937365

**Comment by hewison-chris** [comment]
I don't see any new logging. Did you think it will be too much logging?

**Comment by omerfirmak** [comment]
> I don't see any new logging. Did you think it will be too much logging?

Check L536 and L880. I've added only to the places you commented on.

**Comment by hewison-chris** [comment]
> @hewison-chris , is this the failure we were seeing in NetworkManager tests?
> https://github.com/bpfkorea/agora/pull/1495/checks?check_run_id=1739937365

Sorry, looks like the new network unit test for PeriodicCatchup can sometimes fail. Will take a look now.

File: source/agora/consensus/protocol/Nominator.d
Comment (hewison-chris) [inline source/agora/consensus/protocol/Nominator.d:735]: Also needed here.

File: source/agora/node/Ledger.d
Comment (hewison-chris) [inline source/agora/node/Ledger.d:247]: Also here

File: source/agora/node/Ledger.d
Comment (hewison-chris) [inline source/agora/node/Ledger.d:274]: Also here

**Review by hewison-chris — APPROVED** [review]
LGTM. Just a couple more log lines.

**Comment by omerfirmak** [comment]
Added more logs. 

`PeriodicCatchup` is still failing though. Should we wait for you to take a look @hewison-chris ?

**Comment by hewison-chris** [comment]
> Added more logs.
> 
> `PeriodicCatchup` is still failing though. Should we wait for you to take a look @hewison-chris ?

Yes I know why it fails. Fixing it now. I will have to gossip the signatures also after receiving block during catchup.

**Pull request merged**
